r/tests/testthat/test-dataset.R

```diff
@@ -1778,3 +1778,60 @@ test_that("Collecting zero columns from a dataset doesn't return entire dataset"
     c(32, 0)
   )
 })
+
+# see https://issues.apache.org/jira/browse/ARROW-12315
+test_that("Max partitions fails with non-integer values and less than required partitions values", {
+  skip_if_not_available("parquet")
+  tmp <- tempfile()
+
+  # this example needs 3 partitions
+
+  # max_partitions = chr => error
+  expect_error(
+    mtcars %>%
+      group_by(cyl) %>%
+      write_dataset(tmp, format = "parquet", max_partitions = "foobar")
+  )
```

Comment:
We should assert what each of these errors contain. We don't need to do the full thing, but let's make sure that they are erroring with something useful about partitions

Reply:
in this other case, the error is 
```
 Error: Expected single integer value 
```

Reply:
And like I mention below: if we are using the error handling from c++, we don't need to test it in R as well if it has tests in c++ and there's no other indirection going on.